proposals/0002-monitoring.md

```diff
@@ -0,0 +1,178 @@
+
+This proposal does not talk about health checks. There will be a separate proposal.
+
+Mailinglist thread: [Monitoring and Healthcheck requirements](https://groups.google.com/forum/#!topic/microprofile/jIZAKiu76ys)
+
+## Motivation
+
+Reliable service of a platform needs monitoring. There is already JMX,
+but remote-JMX is not easy to deal with and especially does not fit well
+in a polyglott environment where other services are not running on the 
+JVM.
+To enable monitoring in an easy fashion it is necessary that all microprofile
+implementations follow a certain standard with respect to (base) API path,
+data types involved and return codes used.
+
+## Proposed solution
+
+Data is exposed via REST over http under the `/metrics` base path
+
+* Metrics will respond to GET requests and are exposed in a tree like fashion with sub-trees for various cases described below.
+* A 'shadow tree' that responds to OPTIONS will provide the metadata                 
+* All data is at least exposed as application/Json data type.
+* The api is discoverable. GETting a base object will list references to children
+
+Data access must honour the http response codes, especially
+
+* 200 for successful retrieval of an object
+* 204 when retrieving a subtree that would exist, but has no content. E.g. when the application-specific subtree has no application specific metrics defined.
+* 404 if an directly-addressed item does not exist. This may be a non-existing sub-tree or non-existing object
+* 503 to indicate that a health check resulted in "bad health". The body SHOULD contain details { "details": <text> }
+
+The Api MUST NOT return a 500 Internal Server Error code to represent a non-existing resource or to indicate that a health-check resulted in "bad health"
+
+## Detailed design (if applicable)
+
+### Api Objects
+
+Api-objects MAY include one or more metrics as in 
+
+
+    {
+      "thread-count" : 33,
+      "peak-thread-count" : 47,
+      "total-started-thread-count" : 49,
+      "current-thread-cpu-time" : 66855000,
+      "current-thread-user-time" : 64003000
+    }
+
+or
+
+    {
+      "count": 45
+    }  
+
+### Metadata
+
+Metric values MUST have metadata 
+
+* unit: a fixed set of string units from e.g. [1], [UoM] or [Metrics2.0]  
+* type: 
+     * counter: a monotonically increasing or decreasing numeric value (e.g. total number of requests received)
+     * gauge: a numeric value that can arbitrarily go up and down
+     * bool: a boolean value which can be `true` or `false`
+     * string: a string
+* description (optional): A human readable description of the metric
+* displayName (optional): A human readable name of the metric for display purposes if the metric name is not
+human readable. This could e.g. be the case when the metric name is a uuid.
+* tags (optional): A list of key=value pairs, which are separated by comma. 
+
+Example: 
+
+if `GET /metrics/foo` exposes:
+
+      {"fooVal": 12345} 
+
+then `OPTIONS /metrics/foo` will expose:
```

Comment:
How will OPTIONS work if we have multiple values like the example provided?
```
{
      "thread-count" : 33,
       "peak-thread-count" : 47,
       "total-started-thread-count" : 49,
       "current-thread-cpu-time" : 66855000,
       "current-thread-user-time" : 64003000
}
```

Like this?
``` 
 {
        "thread-count": {
           "unit": "u",
           "type": "counter",
           "description": "",
           "displayName": "",
           "tags": "app=thread"
          },        
        "peak-thread-count":{
           "unit": "u",
           "type": "counter",
           "description": "",
           "displayName": "",
           "tags": "app=thread"
        }
        ...
} 
```

I think would be good to have an example just to be clear.